File: samples/Java/deal-cards-for-freecell.java
import java.util.Arrays;

public class Shuffler {
	
	private int seed;
	
	private String[] deck = {
			"AC", "AD", "AH", "AS",
			"2C", "2D", "2H", "AS",
			"3C", "3D", "3H", "3S",
			"4C", "4D", "4H", "4S",
			"5C", "5D", "5H", "5S",
			"6C", "6D", "6H", "6S",
			"7C", "7D", "7H", "7S",
			"8C", "8D", "8H", "8S",
			"9C", "9D", "9H", "9S",
			"TC", "TD", "TH", "TS",
			"JC", "JD", "JH", "JS",
			"QC", "QD", "QH", "QS",
			"KC", "KD", "KH", "KS",
	};
	
	private int random() {
		seed = (214013 * seed + 2531011) & Integer.MAX_VALUE;
		return seed >> 16;
	}
	
	//shuffled cards go to the end
	private String[] getShuffledDeck() {
		String[] deck = Arrays.copyOf(this.deck, this.deck.length);
		for(int i = deck.length - 1; i > 0; i--) {
			int r = random() % (i + 1);
			String card = deck[r];
			deck[r] = deck[i];
			deck[i] = card;
		}
		return deck;
	}
	
	//deal from end first
	public void dealGame(int seed) {
		this.seed = seed;
		String[] shuffledDeck = getShuffledDeck();
		for(int count = 1, i = shuffledDeck.length - 1; i >= 0; count++, i--) {
			System.out.print(shuffledDeck[i]);
			if(count % 8 == 0) {
				System.out.println();
			} else {
				System.out.print(" ");
			}
		}
		System.out.println();
	}
	
	public static void main(String[] args) {
		Shuffler s = new Shuffler();
		s.dealGame(1);
		System.out.println();
		s.dealGame(617);
	}
	
}

File: samples/Ruby/deal-cards-for-freecell.rb
# Deal cards for FreeCell.
# http://rosettacode.org/wiki/Deal_cards_for_FreeCell

require 'optparse'

# Parse command-line arguments.
#   games = ARGV converted to Integer
#   No arguments? Pick any of first 32000 games.
games = nil
OptionParser.new do |o|
  begin
    o.banner = "Usage: #{o.program_name} number..."
    o.parse!
    games = ARGV.map {|s| Integer(s)}
    games.empty? and games = [rand(32000)]
  rescue => e
    $stderr.puts e, o
    abort
  end
end

# Define methods for old Ruby versions.
#   Enumerable#each_slice appeared in Ruby 1.8.7.
#   Enumerable#flat_map appeared in Ruby 1.9.2.
module Enumerable
  unless method_defined? :each_slice
    def each_slice(count)
      block_given? or return enum_for(:each_slice, count)
      ary = []
      each {|e|
        ary << e
        ary.length == count and (yield ary.dup; ary.clear)}
      ary.empty? or yield ary.dup
      nil
    end
  end

  unless method_defined? :flat_map
    def flat_map
      block_given? or return enum_for(:flat_map)
      ary = []
      each {|e|
        y = yield e
        ary.concat(y) rescue ary.push(y)}
      ary
    end
  end
end

# Create original deck of 52 cards, not yet shuffled.
orig_deck = %w{A 2 3 4 5 6 7 8 9 T J Q K
}.flat_map {|rank| %w{C D H S}.map {|suit| "#{rank}#{suit}"}}

games.each do |seed|
  deck = orig_deck.dup

  # Shuffle deck with random index from linear congruential
  # generator like Microsoft.
  state = seed
  52.downto(2) do |len|
    state = ((214013 * state) + 2531011) & 0x7fff_ffff
    index = (state >> 16) % len
    last = len - 1
    deck[index], deck[last] = deck[last], deck[index]
  end

  deck.reverse!  # Shuffle did reverse deck. Do reverse again.

  # Deal cards.
  puts "Game ##{seed}"
  deck.each_slice(8) {|row| puts " " + row.join(" ")}
end
